File: python/largest-series-product/largest_series_product.py
from functools import reduce

def largest_product(number_string, nums):
  length = len(number_string) - nums + 1
  if length <= 0 or nums < 0:
    raise ValueError
  if nums == 0:
    return 1

  products = []

  for i in range(length):
    sub_string = number_string[i:i+nums]

    int_list = [int(x) for x in sub_string]
    product = reduce(lambda x, acc: x*acc, int_list)
    products.append(product)

  return max(products)
  

File: elixir/largest-series-product/largest_series_product.exs
defmodule Series do

  @doc """
  Finds the largest product of a given number of consecutive numbers in a given string of numbers.
  """
  @spec largest_product(String.t, non_neg_integer) :: non_neg_integer
  def largest_product(_number_string, 0), do: 1
  def largest_product(number_string, size) when size > 0 do
    number_string
    |> String.codepoints
    |> Enum.map(&String.to_integer/1)
    |> Enum.chunk(size, 1)
    |> Enum.map(&product/1)
    |> Enum.max(fn -> raise ArgumentError end)
  end

  def largest_product(_number_string, _size) do
    raise ArgumentError
  end

  defp product(list) do
    Enum.reduce(list, 1, &(*/2))
  end
end
